File: backend/flask/app/services/auth_service.py
from __future__ import annotations

from typing import Any

import bcrypt
from flask_jwt_extended import create_access_token
from werkzeug.security import check_password_hash, generate_password_hash

from app.extensions import db
from app.models import User


def register_user(email: str, password: str, name: str) -> tuple[dict[str, Any], dict[str, Any]]:
    if User.query.filter_by(email=email.lower()).first():
        raise ValueError("Email already registered")
    u = User(
        email=email.lower(),
        password_hash=generate_password_hash(password),
        name=name,
    )
    db.session.add(u)
    db.session.commit()
    tokens = _tokens(u)
    return _user_dict(u), tokens


def login_user(email: str, password: str) -> tuple[dict[str, Any], dict[str, Any]]:
    u = User.query.filter_by(email=email.lower()).first()
    if not u or not _verify_password(u.password_hash, password):
        raise PermissionError("Invalid credentials")
    return _user_dict(u), _tokens(u)


def get_user_by_id(user_id: int) -> dict[str, Any] | None:
    u = User.query.get(user_id)
    return _user_dict(u) if u else None


def _user_dict(u: User) -> dict[str, Any]:
    return {
        "id": u.id,
        "email": u.email,
        "name": u.name,
        "created_at": u.created_at.isoformat() + "Z",
    }


def _tokens(u: User) -> dict[str, Any]:
    token = create_access_token(identity=str(u.id))
    return {"access_token": token, "token_type": "Bearer"}


def _verify_password(stored_hash: str, password: str) -> bool:
    # For users created by Flask itself.
    try:
        if check_password_hash(stored_hash, password):
            return True
    except ValueError:
        # Non-werkzeug format (e.g. seed bcrypt hash) should fall through.
        pass

    # For MySQL seed hash used by Laravel-style bcrypt ($2y$...).
    if stored_hash.startswith("$2y$") or stored_hash.startswith("$2b$"):
        normalized = stored_hash.replace("$2y$", "$2b$", 1)
        return bcrypt.checkpw(password.encode("utf-8"), normalized.encode("utf-8"))

    return False

File: backend/laravel/app/Services/AuthService.php
<?php

namespace App\Services;

use App\Models\User;
use Illuminate\Validation\ValidationException;
use Tymon\JWTAuth\Facades\JWTAuth;

class AuthService
{
    public function register(string $email, string $password, string $name): array
    {
        $email = strtolower($email);
        if (User::where('email', $email)->exists()) {
            throw ValidationException::withMessages(['email' => ['Email already registered']]);
        }

        $user = User::create([
            'email' => $email,
            'password' => $password,
            'name' => $name,
        ]);

        $token = JWTAuth::fromUser($user);

        return [
            'user' => $this->formatUser($user),
            'tokens' => [
                'access_token' => $token,
                'token_type' => 'Bearer',
            ],
        ];
    }

    public function login(string $email, string $password): array
    {
        $email = strtolower($email);
        $token = JWTAuth::attempt(['email' => $email, 'password' => $password]);
        if (! $token) {
            throw ValidationException::withMessages(['email' => ['Invalid credentials']]);
        }

        $user = User::where('email', $email)->firstOrFail();

        return [
            'user' => $this->formatUser($user),
            'tokens' => [
                'access_token' => $token,
                'token_type' => 'Bearer',
            ],
        ];
    }

    private function formatUser(User $user): array
    {
        return [
            'id' => $user->id,
            'email' => $user->email,
            'name' => $user->name,
            'created_at' => $user->created_at?->toIso8601String(),
        ];
    }
}
